File: LivrusVirtus/BoaConstructor/wxPython/TextValidator.py
from wxPython.wx import *
import string

class TextValidator(wxPyValidator):
  FILTER_NONE = 0
  FILTER_ALPHA = 1
  FILTER_NUMERIC = 2

  def __init__(self, flag=None, pyVar=None):
    wxPyValidator.__init__(self)
    self.flag = flag
    EVT_CHAR(self, self.OnChar)

  def Clone(self):
    return TextValidator(self.flag)

  def Validate(self, win):
    tc = self.GetWindow()
    val = tc.GetValue()
    if self.flag == self.FILTER_ALPHA:
      for x in val:
        if x not in string.letters:
          return False

    elif self.flag == self.FILTER_NUMERIC:
      for x in val:
        if x not in string.digits:
          return False

    elif self.flag == self.FILTER_NONE:
      return True

    return True

  def OnChar(self, event):
    if self.flag != self.FILTER_NONE:
      key = event.KeyCode()
      if key < WXK_SPACE or key == WXK_DELETE or key > 255:
        event.Skip()
        return
      if self.flag == self.FILTER_ALPHA and chr(key) in string.letters:
        event.Skip()
        return
      if self.flag == self.FILTER_NUMERIC and chr(key) in string.digits:
        event.Skip()
        return
  
      if not wxValidator_IsSilent():
        wxBell()

      return

    else:
      event.Skip()

File: LivrusVirtus/wxWindows/wxJava/src/Livrus/TextValidator.java
package Livrus;

import org.wxwindows.wxKeyEvent;
import org.wxwindows.wxKeyListener;
import org.wxwindows.wxObject;
import org.wxwindows.wxValidator;
import org.wxwindows.wxWindow;
import org.wxwindows.wxTextCtrl;

public class TextValidator
    extends wxValidator {
  public static int FILTER_NONE = 0;
  public static int FILTER_ALPHA = 1;
  public static int FILTER_NUMERIC = 2;
  private int nLocalFiltro, nLocalMaxTam, nTam;
  private wxTextCtrl edtLocal;

  public TextValidator(wxTextCtrl edt, int nfiltro, int nMaxTam) {
    super();
    edtLocal = edt;
    nLocalFiltro = nfiltro;
    nLocalMaxTam = nMaxTam;
    nTam = 0;
    EVT_CHAR(new OnChar());
  }

  public wxObject Clone() {
    return new TextValidator(edtLocal, nLocalFiltro, nLocalMaxTam);
  }

  public boolean Validate(wxWindow window) {
    return false;
  }

  private class OnChar
      implements wxKeyListener {
    public void handleEvent(wxKeyEvent event) {
        int key = event.GetKeyCode();
        nTam = edtLocal.GetValue().length();
        if ( (key == WXK_LEFT) || (key == WXK_RIGHT) || (key == WXK_HOME) ||
             (key == WXK_END)) {
          event.Skip();
          return;
        }
        if ( (key == WXK_BACK) || (key == WXK_DELETE)) {
          if (nTam > 0)
            nTam--;
          event.Skip();
          return;
        }
        if ( (key < WXK_SPACE) || (key > 255)) {
          if (nTam < nLocalMaxTam){
            nTam++;
            event.Skip();
          }
          return;
        }
      if (nLocalFiltro != FILTER_NONE) {
        char ch = (char) key;
        if ( (nLocalFiltro == FILTER_ALPHA) && Character.isLetter(ch)) {
          if (nTam < nLocalMaxTam){
            nTam++;
            event.Skip();
          }
          return;
        }
        if ( (nLocalFiltro == FILTER_NUMERIC) && Character.isDigit(ch)) {
          if (nTam < nLocalMaxTam){
            nTam++;
            event.Skip();
          }
          return;
        }
        if (!wxValidator.IsSilent()) {
          wxBell();
        }
        return;
      } else {
        if (nTam < nLocalMaxTam){
          nTam++;
          event.Skip();
        }
        return;
      }

    }
  }

  public boolean TransferFromWindow() {
    return true;
  }

  public boolean TransferToWindow() {
    return true;
  }

}
